diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index db516a2..a3e538c 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -333,8 +333,8 @@ LIBDIR={- our $libdir = $config{libdir};
 # $(libdir) is chosen to be compatible with the GNU coding standards
 libdir={- file_name_is_absolute($libdir)
           ? $libdir : '$(INSTALLTOP)/$(LIBDIR)' -}
-ENGINESDIR=$(libdir)/engines-{- $sover_dirname -}
-MODULESDIR=$(libdir)/ossl-modules
+ENGINESDIR=$(libdir)/openssl/engines-{- $sover_dirname -}
+MODULESDIR=$(libdir)/openssl/ossl-modules
 
 # Convenience variable for those who want to set the rpath in shared
 # libraries and applications
