diff --git a/crypto/dllmain.c b/crypto/dllmain.c
index 1525878..e7ef45f 100644
--- a/crypto/dllmain.c
+++ b/crypto/dllmain.c
@@ -35,7 +35,9 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
     case DLL_THREAD_ATTACH:
         break;
     case DLL_THREAD_DETACH:
+# ifndef __CYGWIN__
         OPENSSL_thread_stop();
+# endif
         break;
     case DLL_PROCESS_DETACH:
         break;
diff --git a/providers/fips/self_test.c b/providers/fips/self_test.c
index 008a4fa..fa5dc9a 100644
--- a/providers/fips/self_test.c
+++ b/providers/fips/self_test.c
@@ -101,7 +101,9 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
         init();
         break;
     case DLL_PROCESS_DETACH:
+# ifndef __CYGWIN__
         cleanup();
+# endif
         break;
     default:
         break;
