diff --git a/examples/complete/bash_completion b/examples/complete/bash_completion
index b0cf4a8..1362eaa 100644
--- a/examples/complete/bash_completion
+++ b/examples/complete/bash_completion
@@ -48,6 +48,7 @@ readonly BASH_COMPLETION BASH_COMPLETION_DIR
 UNAME=$( uname -s )
 # strip OS type and version under Cygwin (e.g. CYGWIN_NT-5.1 => Cygwin)
 UNAME=${UNAME/CYGWIN_*/Cygwin}
+UNAME=${UNAME/MSYS_*/Msys}
 RELEASE=$( uname -r )
 
 # features supported by bash 2.05 and higher
@@ -817,7 +818,7 @@ complete -F _insmod $filenames insmod modprobe modinfo
 #
 [ $UNAME = GNU -o $UNAME = Linux -o $UNAME = Darwin \
   -o $UNAME = FreeBSD -o $UNAME = SunOS -o $UNAME = Cygwin \
-  -o $UNAME = OpenBSD ] &&
+  -o $UNAME = Msys -o $UNAME = OpenBSD ] &&
 _man()
 {
 	local cur prev sect manpath UNAME
@@ -837,8 +838,9 @@ _man()
 	UNAME=$( uname -s )
 	# strip OS type and version under Cygwin
 	UNAME=${UNAME/CYGWIN_*/Cygwin}
+	UNAME=${UNAME/MSYS_*/Msys}
 	if [ $UNAME = GNU -o $UNAME = Linux -o $UNAME = FreeBSD \
-	     -o $UNAME = Cygwin ]; then
+	     -o $UNAME = Cygwin -o $UNAME = Msys ]; then
 		manpath=$( manpath 2>/dev/null || command man --path )
 	else
 		manpath=$MANPATH
@@ -873,7 +875,7 @@ _man()
 }
 [ $UNAME = GNU -o $UNAME = Linux -o $UNAME = Darwin \
   -o $UNAME = FreeBSD -o $UNAME = SunOS -o $UNAME = Cygwin \
-  -o $UNAME = OpenBSD ] && \
+  -o $UNAME = Msys -o $UNAME = OpenBSD ] && \
 complete -F _man $filenames man apropos whatis
 
 # renice(8) completion
@@ -4264,7 +4266,7 @@ _gcc()
 	fi
 } &&
 complete $filenames -F _gcc gcc g++ c++ g77 gcj gpc
-[ $UNAME = GNU -o $UNAME = Linux -o $UNAME = Cygwin ] && \
+[ $UNAME = GNU -o $UNAME = Linux -o $UNAME = Cygwin -o $UNAME = Msys ] && \
 [ -n "${have:-}" ] && complete $filenames -F _gcc cc
 
 # Linux cardctl(8) completion
diff --git a/support/config.rpath b/support/config.rpath
index fc5913d..54df920 100755
--- a/support/config.rpath
+++ b/support/config.rpath
@@ -57,7 +57,7 @@ else
     aix*)
       wl='-Wl,'
       ;;
-    mingw* | cygwin* | pw32* | os2* | cegcc*)
+    mingw* | cygwin* | msys* | pw32* | os2* | cegcc*)
       ;;
     hpux9* | hpux10* | hpux11*)
       wl='-Wl,'
@@ -149,7 +149,7 @@ hardcode_direct=no
 hardcode_minus_L=no
 
 case "$host_os" in
-  cygwin* | mingw* | pw32* | cegcc*)
+  cygwin* | msys* | mingw* | pw32* | cegcc*)
     # FIXME: the MSVC++ port hasn't been tested in a loooong time
     # When not using gcc, we currently assume that we are using
     # Microsoft Visual C++.
@@ -198,7 +198,7 @@ if test "$with_gnu_ld" = yes; then
         ld_shlibs=no
       fi
       ;;
-    cygwin* | mingw* | pw32* | cegcc*)
+    cygwin* | msys* | mingw* | pw32* | cegcc*)
       # hardcode_libdir_flag_spec is actually meaningless, as there is
       # no search path for DLLs.
       hardcode_libdir_flag_spec='-L$libdir'
@@ -348,7 +348,7 @@ else
       ;;
     bsdi[45]*)
       ;;
-    cygwin* | mingw* | pw32* | cegcc*)
+    cygwin* | msys* | mingw* | pw32* | cegcc*)
       # When not using gcc, we currently assume that we are using
       # Microsoft Visual C++.
       # hardcode_libdir_flag_spec is actually meaningless, as there is
@@ -533,7 +533,7 @@ case "$host_os" in
   bsdi[45]*)
     library_names_spec='$libname$shrext'
     ;;
-  cygwin* | mingw* | pw32* | cegcc*)
+  cygwin* | msys* | mingw* | pw32* | cegcc*)
     shrext=.dll
     library_names_spec='$libname.dll.a $libname.lib'
     ;;
diff --git a/support/shobj-conf b/support/shobj-conf
index cd7634d..6bd693c 100755
--- a/support/shobj-conf
+++ b/support/shobj-conf
@@ -484,6 +484,23 @@ cygwin*)
 	SHLIB_LIBVERSION='$(SHLIB_DLLVERSION).$(SHLIB_LIBSUFF)'
 	SHLIB_LIBS='$(TERMCAP_LIB)'
 
+	SHLIB_DOT=
+	# For official cygwin releases, DLLVERSION will be defined in the
+	# environment of configure, and will be incremented any time the API
+	# changes in a non-backwards compatible manner.  Otherwise, it is just
+	# SHLIB_MAJOR.
+	if [ -n "$DLLVERSION" ] ; then
+		SHLIB_DLLVERSION="$DLLVERSION"
+	fi
+	;;
+msys*)
+	SHOBJ_LD='$(CC)'
+	SHOBJ_LDFLAGS='-shared -Wl,--enable-auto-import -Wl,--enable-auto-image-base -Wl,--export-all -Wl,--out-implib=$(@).a'
+	SHLIB_LIBPREF='msys-'
+	SHLIB_LIBSUFF='dll'
+	SHLIB_LIBVERSION='$(SHLIB_DLLVERSION).$(SHLIB_LIBSUFF)'
+	SHLIB_LIBS='$(TERMCAP_LIB)'
+
 	SHLIB_DOT=
 	# For official cygwin releases, DLLVERSION will be defined in the
 	# environment of configure, and will be incremented any time the API
